roll/parser.py

```
@@ -0,0 +1,234 @@
+    primary = number | string | bracketed | let_stmt | anon_func | variable
+    bracketed = "(" >> expr << ")"
+
+    func = identifier & func_decl > func_decl_or_type  # (func_decl | type_decl)
+    func_decl = "=" >> expr  # expr << "=" & expr
+    type_decl = "::" >> type_sig
+    type_sig = rep1(type_brac | type_num | type_string)
+    type_brac = "(" >> type_sig << ")"
+    type_num = "#"
+    type_string = "$"
+
+    program = repsep("@" >> func | expr, ";") << opt(";") > Program
+
+    main = program
+
+
+# def remove_comments(string):
+#     return re.sub(r"\/\*(?s).*\*\/|//.*", "", string)
+
+
+class DiscordParser(TextParsers):
+    """Removes surrounding code blocks before the program can reach the main parser"""
+
+    main = (
+        "```" >> reg(r"(?s).*?(?=```)") << "```"
+        | "`" >> reg(r"(?s).*?(?=`)") << "`"
+        | reg(r"[^`](?s).*")
+    )
+
+
+def parse_program(source: str):
+    try:
+        no_blocks = DiscordParser.main.parse(source).or_die()
+    except ParseError:
+        raise ParseError("Unclosed code blocks")
+    try:
+        ast = ProgramParser.main.parse(no_blocks).or_die()
+    except ParseError as e:
+        raise ParseError(format_parse_error(e, source))
+    return ast
+
+
+def format_parse_error(err, source):
+    found = re.search(r"(?<=but found ').*?(?=')", err.message)
+    if found is None:
+        last_line = source.split("\n")[-1]
+        pointer = last_line + "\n" + " " * (len(last_line) - 1) + "^"
+        return f"Found unexpected end of source\n{pointer}"
+    else:
+        found = found.group(0)
+        try:
+            line = re.findall(r"(?<=Line )\d+", err.message)[-1]
+            char = re.findall(r"(?<=character )\d+", err.message)[-1]
+            pointer = (
+                source.split("\n")[int(line) - 1] + "\n" + " " * (int(char) - 1) + "^"
+            )
+            return f"Unexpected token {found}\nLine: {line}\nChar: {char}\n{pointer}"
+        except:

```

Bare except, needs refining.